Firebase/Auth/Source/FIRAuth.m

```diff
@@ -512,46 +512,59 @@ - (void)signInWithEmail:(NSString *)email
                password:(NSString *)password
              completion:(FIRAuthResultCallback)completion {
   dispatch_async(FIRAuthGlobalWorkQueue(), ^{
-    [self signInWithEmail:email
-                 password:password
-                 callback:[self signInFlowAuthResultCallbackByDecoratingCallback:completion]];
-  });
-}
-
-/** @fn signInWithEmail:password:callback:
-    @brief Signs in using an email address and password.
-    @param email The user's email address.
-    @param password The user's password.
-    @param callback A block which is invoked when the sign in finishes (or is cancelled.) Invoked
-        asynchronously on the global auth work queue in the future.
-    @remarks This is the internal counterpart of this method, which uses a callback that does not
-        update the current user.
- */
-- (void)signInWithEmail:(NSString *)email
-               password:(NSString *)password
-               callback:(FIRAuthResultCallback)callback {
-  FIRVerifyPasswordRequest *request =
-      [[FIRVerifyPasswordRequest alloc] initWithEmail:email
-                                             password:password
-                                 requestConfiguration:_requestConfiguration];
-
-  if (![request.password length]) {
-    callback(nil, [FIRAuthErrorUtils wrongPasswordErrorWithMessage:nil]);
-    return;
-  }
-  [FIRAuthBackend verifyPassword:request
-                        callback:^(FIRVerifyPasswordResponse *_Nullable response,
-                                   NSError *_Nullable error) {
-    if (error) {
-      callback(nil, error);
-      return;
-    }
-    [self completeSignInWithAccessToken:response.IDToken
+    FIRAuthResultCallback decoratedCallback =
+        [self signInFlowAuthResultCallbackByDecoratingCallback:completion];
+    [self internalSignInWithEmail:email
+                         password:password
+                       completion:^(FIRVerifyPasswordResponse *_Nullable response,
+                                    NSError *_Nullable error) {
+      if (error) {
+        decoratedCallback(nil, error);
+        return;
+      }
+      [self completeSignInWithAccessToken:response.IDToken
               accessTokenExpirationDate:response.approximateExpirationDate
                            refreshToken:response.refreshToken
                               anonymous:NO
-                               callback:callback];
-  }];
+                               callback:decoratedCallback];
+    }];
+  });
+}
+
+- (void)signInAndRetrieveDataWithEmail:(NSString *)email
+                              password:(NSString *)password
+                            completion:(FIRAuthDataResultCallback)completion {
+  dispatch_async(FIRAuthGlobalWorkQueue(), ^{
```

Comment:
Same question here: can we call `signInAndRetriveDataWithCredential:completion:` to avoid repeated logic?

Reply:
Done